discoverd2/server/state.go

```diff
@@ -0,0 +1,332 @@
+package server
+
+import (
+	"container/list"
+	"errors"
+	"sync"
+
+	"github.com/flynn/flynn/pkg/stream"
+)
+
+type EventKind uint
+
+const (
+	EventKindUp EventKind = 1 << iota
+	EventKindUpdate
+	EventKindDown
+	EventKindLeader
+)
+
+func (k EventKind) String() string {
+	switch k {
+	case EventKindUp:
+		return "up"
+	case EventKindUpdate:
+		return "update"
+	case EventKindDown:
+		return "down"
+	case EventKindLeader:
+		return "leader"
+	default:
+		return "unknown"
+	}
+}
+
+type Event struct {
+	Service string
+	Kind    EventKind
+	*Instance
+}
+
+func eventKindUpdate(existing bool) EventKind {
+	if existing {
+		return EventKindUpdate
+	}
+	return EventKindUp
+}
+
+// Instance is a single running instance of a service.
+type Instance struct {
+	// ID is unique within the service, and is currently defined as
+	// Hex(SHA256(Proto + "-" + Addr)) but this may change in the future.
+	ID string `json:"id"`
+
+	// Addr is the IP/port address that can be used to communicate with the
+	// service. It must be valid to dial this address.
+	Addr string `json:"addr"`
+
+	// Proto is the protocol used to connect to the service, examples include:
+	// tcp, udp, http, https. It must be lowercase alphanumeric.
+	Proto string `json:"proto"`
+
+	// Meta is arbitrary metadata specified when registering the instance.
+	Meta map[string]string `json:"meta,omitempty"`
+
+	// Leader is true if the instance is the leader of the service. Exactly one
+	// instance per service has this set to true at any point in time.
+	Leader bool `json:"leader,omitempty"`
+
+	// Index is the logical epoch of the initial registration of the instance.
```

Comment:
Should this read something like:

```
Index is the number of registrations since the initial registration of the service
```

If not, maybe I'm misunderstanding what this is.